api/src/main/java/net/kyori/adventure/text/event/ClickCallback.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * This file is part of adventure, licensed under the MIT License.
+ *
+ * Copyright (c) 2017-2023 KyoriPowered
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.event;
+
+import java.time.Duration;
+import java.time.temporal.TemporalAmount;
+import java.util.function.Predicate;
+import net.kyori.adventure.audience.Audience;
+import net.kyori.adventure.builder.AbstractBuilder;
+import net.kyori.adventure.permission.PermissionChecker;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.util.PlatformAPI;
+import net.kyori.examination.Examinable;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A handler for callback click events.
+ *
+ * @param <T> audience type
+ * @since 4.13.0
+ */
+@FunctionalInterface
+public interface ClickCallback<T extends Audience> {
+  /**
+   * Adjust this callback to accept any audience, and perform the appropriate filtering.
+   *
+   * @param <W> the wider type
+   * @param <N> the narrower type
+   * @param original the original callback of a narrower audience type
+   * @param type the audience type to accept
+   * @param failureMessage the message to send to the audience if it is not of the appropriate type
+   * @return a new callback
+   * @since 4.13.0
+   */
+  static <W extends Audience, N extends W> @NotNull ClickCallback<W> widen(final @NotNull ClickCallback<N> original, final @NotNull Class<N> type, final @Nullable Component failureMessage) {
+    return audience -> {
+      if (type.isInstance(audience)) {
+        original.accept(type.cast(audience));
+      } else if (failureMessage != null) {
+        audience.sendMessage(failureMessage);
```

Comment:
Perhaps instead of having a `Component` argument here, we could have a `Consumer<? super Audience>` instead?
```suggestion
  static <W extends Audience, N extends W> @NotNull ClickCallback<W> widen(final @NotNull ClickCallback<N> original, final @NotNull Class<N> type, final @Nullable Consumer<? super Audience> onTypeMismatch) {
    return audience -> {
      if (type.isInstance(audience)) {
        original.accept(type.cast(audience));
      } else if (onTypeMismatch != null) {
        onTypeMismatch.accept(audience);
```